src/core.jl

```diff
@@ -1,35 +1,86 @@
+"""
+    JSONRPCError
+
+An object representing a JSON-RPC Error.
+
+Fields:
+ * code::Int
+ * msg::AbstractString
+ * data::Any
+
+See Section 5.1 of the JSON RPC 2.0 specification for more information.
+"""
 struct JSONRPCError <: Exception
     code::Int
     msg::AbstractString
     data::Any
 end
 
+"""
+    PARSE_ERROR
+
+Invalid JSON was received by the server.
+An error occurred on the server while parsing the JSON text.
+"""
+const PARSE_ERROR = -32700
+
+"""
+    INVALID_REQUEST
+
+The JSON sent is not a valid Request object.
+"""
+const INVALID_REQUEST = -32600
+
+"""
+    METHOD_NOT_FOUND
+
+The method does not exist / is not available.
+"""
+const METHOD_NOT_FOUND = -32601
+
+"""
+    INVALID_PARAMS
+
+Invalid method parameter(s).
+"""
+const INVALID_PARAMS = -32602
+
+"""
+    INTERNAL_ERROR
+
+Internal JSON-RPC error.
+"""
+const INTERNAL_ERROR = -32603
+
+# these are the reserved endpoints of JSON-RPC error codes
+const SERVER_ERROR_END       = -32000
+const SERVER_ERROR_START     = -32099
+
+"""
+   RPCErrorStrings
+
+A `Base.IdDict` containing the mapping of JSON-RPC error codes to a short, descriptive string.
+
+Use this to hook into `showerror(io::IO, ::JSONRPCError)` for display purposes. A default fallback to `"Unknown"` exists.
+"""
+const RPCErrorStrings = Base.IdDict(
+    PARSE_ERROR => "ParseError",
+    INVALID_REQUEST => "InvalidRequest",
+    METHOD_NOT_FOUND => "MethodNotFound",
+    INVALID_PARAMS => "InvalidParams",
+    INTERNAL_ERROR => "InternalError",
+    [ i => "ServerError" for i in SERVER_ERROR_START:SERVER_ERROR_END]...,
+    # TODO: these should be removed - they are in the application/implementation specific range
+    -32000 => "serverErrorEnd",
+    -32099 => "serverErrorStart",
+    -32002 => "ServerNotInitialized",
+    -32001 => "UnknownErrorCode",
+    -32800 => "RequestCancelled",
+	-32801 => "ContentModified"
```

Comment:
Not sure what you mean by that. These are implementation specific -- this in fact is the implementation they are specified by.

Reply:
Yeah, I'll have to update these a bit - the first two (`serverErrorEnd`/`serverErrorStart`) at least aren't error codes at all, so shouldn't have an error message associated with them by default. The next two are definitely ok, but the last two are (I think) LSP specific, and not necessarily JSONRPC; there isn't necessarily support for cancellable requests or content that would have been modified mid-request (in particular if the underlying implementation is idempotent) by a codebase using this. They're also outside the range of reserved error codes for JSON-RPC servers, so IMO should not live here.